### ISSUE TITLE
Per Package configs

### ISSUE THREAD

**Issue by Morganamilo** [Morganamilo/paru#901]
This is a more complete idea of #188 and #748

The idea is to have add a `packages` or more accurately a  `bases` field.

The env section already exists so with a bases field we could say:

```
# Applies to all packages
[env]
BUILDDIR = foo

# Only applies to spotify and google-chome
[env]
Bases = spotify google-chome
BUILDDIR = foo
PKGDEST = bar
```

Similarly this could also work in options:

```
# Applies to all packages
[options]
MakepkgConf = foo

# Only applies to spotify and google-chome
[options]
Bases = spotify google-chome
MakepkgConf = bar
```

Obviously only certain fields will be overrideable. It doesn't make sense to override IgnoreDevel, NewsOnUpgrade, UpgradeMenu. In fact most fields will not be overrideable which makes me think the section should be called something other than `options`. Maybe `[override]` or `[base]`.

I think a cleaner aproch would to do something like `[env.spotify]` but that would mean you could only do one package at a time and it would lead to duplication as users are likely to want to set common settings for groups of packages.

**Comment by shssoichiro** [comment]
I have a need for this and an idea to implement it. I'll be working on a PR to add it.

**Comment by shssoichiro** [comment]
Here's my proposal, let me know if you have any issues or improvements:

To accomplish this, we will create support for `override` sections. Each `override` section will apply to specific packages defined within it. `options` and `env` sections can support overrides.

Here would be an example config which implements different override profiles
for different packages:

```conf
# Applies to all packages
[options]
MakepkgConf = foo

# Overrides the above, only for spotify and google-chome
[options.override]
Packages = spotify google-chrome
MakepkgConf = bar

# Overrides the default paths for cuda only
[env.override]
Packages = cuda
BUILDDIR = foo
PKGDEST = bar

# make our artificial benchmarks really fast
[env.override]
Packages = pyprime sysbench
CFLAGS = "-Ofast -march=native"
```

Note how we can have multiple override sections. The `Packages` line defines which packages the override applies to. The `Packages` line may appear anywhere within the block; do not assume it is the first line. Any `override` block must have a `Packages` line.

A package may appear in more than one block; the last block of a type for a given package takes precedence.

**Comment by ninetailedtori** [comment]
> Here's my proposal, let me know if you have any issues or improvements:
> 
> To accomplish this, we will create support for `override` sections. Each `override` section will apply to specific packages defined within it. `options` and `env` sections can support overrides.
> 
> Here would be an example config which implements different override profiles for different packages:
> 
> ```toml
> # Applies to all packages
> [options]
> MakepkgConf = foo
> 
> # Overrides the above, only for spotify and google-chome
> [options.override]
> Packages = spotify google-chrome
> MakepkgConf = bar
> 
> # Overrides the default paths for cuda only
> [env.override]
> Packages = cuda
> BUILDDIR = foo
> PKGDEST = bar
> 
> # make our artificial benchmarks really fast
> [env.override]
> Packages = pyprime sysbench
> CFLAGS = "-Ofast -march=native"
> ```
> 
> Note how we can have multiple override sections. The `Packages` line defines which packages the override applies to. The `Packages` line may appear anywhere within the block; do not assume it is the first line. Any `override` block must have a `Packages` line.
> 
> A package may appear in more than one block; the last block of a type for a given package takes precedence.

I quite like this idea, though I had an idea on the naming of categories:
- `[override.package]`
  - for a specific package then, you could do `[override.package.<pkgname>]`
- `[override.pkgbase]`
  - for a specific pkgbase, we then do `[override.pkgbase.<pkgbase>]`
- `[override.group]`
  - `[override.group.<pkggroup>]` which supports groups
  - and `[override.group.custom-<some-group-identifier>]` if defining a custom package group

And I think that would cover everything!

I'd also put the envs to be defined under a nested table, i.e.
```toml
[override.package.pacman-static]
MakepkgConf = "/etc/some-other-makepkg.conf"
Overrides = { CC = "gcc", CXX = "g++", CC_LD = "mold", CXX_LD = "mold" }
```

This would allow some of the lookups on these tables to be much faster, and the table name itself being read then also provides the scope if not a custom group. It also nests all overrides into the table scope `Overrides`, which then mitigates any possible breaking of our own TOML file.

**Comment by ninetailedtori** [comment]
(Alternatively, if you DID want to do your style of arrays, I'd recommend using double square brackets, to use the "array of tables" in TOML as such:
```toml
[options]
MakepkgConf = "/etc/foo.conf"

[[env.override]]
Packages = "spotify google-chrome"
MakepkgConf = "/etc/bar.conf"

[[env.override]]
Packages = "cuda"
Overrides = { BUILDDIR = "/etc/foo/", PKGDEST = "/etc/bar/" }

[[env.override]]
Packages = "pyprime sysbench"
Overrides = { CFLAGS = "-Ofast -march=native" }
```
which would be a lot neater than the original solution to handle imo when parsing the toml itself. I still do have a bias to my own idea though.) :]

**Comment by shssoichiro** [comment]
I also prefer your naming scheme, the issue is as @Morganamilo mentioned, if we tie each override to a specific package, it creates a lot of duplication because you'd likely want several packages to have the same overrides. This is the case for me, since I use clang as my default compiler, but have some packages that only work with gcc, so they require an override. Though I wonder if `MakepkgConf` is already good enough for that use case. 🤔  It does essentially serve as a template definition.

**Comment by shssoichiro** [comment]
Ah, I do see your mention of custom groups. Am I understanding correctly that would solve this use case? So with custom groups, users could do something like this?

```
[override.group.custom-gcc]
Packages = "package1 package2 package3"
Overrides = { CC = "gcc", CXX = "g++", LD = "ld" }
```

**Comment by ninetailedtori** [comment]
Exactly! That'd let us still have groups including custom groups, and it also allows neater labelling of the TOML to the parser imo :3

**Comment by shssoichiro** [comment]
Excellent, I like this. Thanks! Okay, I'll get a PR put together for this soon.
